core/src/apps/cardano/sign_tx.py

```diff
@@ -274,27 +278,15 @@ def _validate_withdrawals(withdrawals: List[CardanoTxWithdrawalType]) -> None:
             raise INVALID_WITHDRAWAL
 
 
-def _validate_metadata(metadata: Optional[bytes]) -> None:
-    if not metadata:
-        return
-
-    if len(metadata) > MAX_METADATA_LENGTH:
-        raise INVALID_METADATA
-
-    try:
-        # this also raises an error if there's some data remaining
-        decoded = cbor.decode(metadata)
-    except Exception:
-        raise INVALID_METADATA
-
-    if not isinstance(decoded, dict):
-        raise INVALID_METADATA
-
-
 def _cborize_signed_tx(
     keychain: seed.Keychain, msg: CardanoSignTx
 ) -> Tuple[CborizedSignedTx, TxHash]:
-    tx_body = _cborize_tx_body(keychain, msg)
+    # we first need to get the cbor encoded auxiliary_data since their hash is used in tx_body
+    auxiliary_data_cbor = get_auxiliary_data_cbor(
+        keychain, msg.auxiliary_data, msg.protocol_magic, msg.network_id
+    )
+
+    tx_body = _cborize_tx_body(keychain, msg, auxiliary_data_cbor)
```

Comment:
can't _cborize_tx_body build the auxiliary data cbor itself - I understand we would be doing it twice and result in additional memory allocation/fragmentation, but I'd not optimize it unless we need to, which I think isn't the case here and rather keep a simpler interface and encapsulate that low-level concern of metadata hash computation

Reply:
Updated - https://github.com/vacuumlabs/trezor-firmware/commit/08c938e56c6a49e9d108095339eec63141bf6799#diff-eeb6dbc0d3a37193fbabe90b03fbdc5ee922ed82b342b639364808bef1f1b4beR150. It is indeed cleaner although there's duplication.